### ISSUE TITLE
Transitive applying of settings files very memory intensive and slow.

### ISSUE THREAD

**Issue by aiosin** [gradle/gradle#7953]
I have a big set of projects (~200) in a flat structure. They were built until now with ant and I have converted them to gradle projects.

Project structure looks like this:
```

root
-A
-B
-C
-D
-E
-F
-G
```

Each project has been transformed to contain a settings.gradle of all dependency projects.
So if a project had project `A` and `B` in its classpath, then the projects settings.gradle will look like this: 

```
apply from :'../A/settings.gradle'
include ':A'
project(':A').projectDir = new File('../A')
apply from :'../B/settings.gradle'
include ':B'
project(':B').projectDir = new File('../B')

```

those projects can too contain settings.gradle files like these.

Currently with some projects (which have more include like these than the others) this takes up and awful amount of time and resources.

If I build my root workspace with a custom settings.gradle where everything is declared right away and gradle does not need to search and compile setting files the build takes ~20 min for all projects, as opposed to when gradle needs to search and apply settings it takes ~10 min for a single(!) project.

### Expected Behavior
I expect it to not eat up all of my RAM.
I expect this to take ~30 sec at most.

### Current Behavior
Currently this eats up all of my ram for some projects, which definetly contain no circular dependencies.
It also takes up to 10 min, while eating up most of my CPU resources for that time. 

### Context

I'm trying to build my project, and need a way to quicken the application for settings.gradle files.

### Steps to Reproduce (for bugs)
I dont have an example project, because the code is not open, and is too big to upload anywhere.

### Your Environment
Happens with gradle 5 and gradle 4.10.*
Windows 10 64 bit.
Installed gradle with chocolatey and interactin with gradle through powershell.

**Comment by big-guy** [comment]
I may be misunderstanding what you're saying, but I think there's a simpler way to do what you're trying to do.

Given this structure, where `A`, `B`, `C`, ... can all have multiple projects in them:
```
root
-A
-B
-C
...
```

If `A` is completely independent from `B`, you could have a separate `settings.gradle` in `A`, `B`, etc, that do not refer to one another (so no `apply from: "../settings.gradle"` sort of thing).  If you still want to tie them all together, say to have a single entry point to import all of the separate builds into an IDE, you can have a `settings.gradle` at the root level that ties all of the independent builds together with `includeBuild(...)`.  You don't need to reference the individual settings.gradle files from separate builds.

If `A`, `B`, `C`, ... can have inter-dependencies on one another, you can do the same as above.  This makes more sense if `A`, `B`, `C`, ... can be released separately or are worked on by separate teams and they integrate together via versioned binary dependencies.

If `A`, `B`, `C`, ... can have inter-dependencies on one another, you can also have a single `settings.gradle` at the root level that references all of the subprojects. 

Bottom line is, don't do what you're doing.
